File: kata/src/day1/Stack.ts
type Node<T> = {
    value: T;
    next: Node<T> | null;
};

export default class Stack<T> {
    public length: number;
    private _head: Node<T> | null;

    constructor() {
        this._head = null;
        this.length = 0;
    }

    push(item: T): void {
        const node: Node<T> = {
            value: item,
            next: null,
        };
        if (this.length === 0) {
            this._head = node;
        } else {
            node.next = this._head;
            this._head = node;
        }
        this.length++;
    }

    pop(): T | undefined {
        if (this.length === 0) {
            return;
        } else if (this.length === 1) {
            const temp = this._head!;
            this._head = this._head!.next;
            this.length = 0;
            return temp.value;
        } else {
            const temp = this._head!;
            this._head = this._head!.next;
            this.length--;
            return temp.value;
        }
    }

    peek(): T | undefined {
        return this._head?.value ?? undefined;
    }
}

File: cpp/src/StacksAndQueues/Stack.cpp
#include "./Stack.h"
#include <iostream>
#include <limits.h>

Stack::Stack(int value) {
    Node* node = new Node(value);
    m_top = node;
    m_height = 1;
}

void Stack::push(int value) {
    Node* node = new Node(value);
    node->next = m_top;
    m_top = node;
    m_height++;
}

int Stack::pop() {
    if (m_height == 0) {
        return INT_MIN;
    } else {
        Node* temp = m_top;
        int poppedValue = m_top->value;
        m_top = m_top->next;
        delete temp;
        m_height--;
        return poppedValue;
    }
}

void Stack::printAll() {
    Node* temp = m_top;
    while (temp) {
        std::cout << temp->value << ' ';
        temp = temp->next;
    }
    std::cout << '\n';
}

void Stack::printTop() const {
    if (!m_top) {
        return;
    }
    std::cout << "TOP:\t" << m_top->value << std::endl;
}

void Stack::printHeight() const {
    std::cout << "HEIGHT:\t" << m_height << std::endl;
}
